### PR TITLE
feat: add scrollBehavior option

### LINKED ISSUE #871: Enable users to change the scrolling strategy
Users have expressed a desire to modify how Cypress internally scrolls elements.

Related to:

- https://github.com/cypress-io/cypress/issues/682
- https://github.com/cypress-io/cypress/issues/483
- https://github.com/cypress-io/cypress/issues/706
- https://github.com/cypress-io/cypress/issues/3623

Currently we pin elements to the topmost coordinate after scrolling.

Users should be able to change this strategy to be one of:

- top
- bottom
- middle (don't scroll if already within screen)
- disable (turn off all automatic scrolling)

### By default, scroll element into center
- [ ] Scrolling strategy make element cannot click #2302

### LINKED ISSUE #4356: option to disable scroll per command (Auto scroll on click/trigger can break virtual grids)
### Current behavior:

Cypress click will scroll an element inside a virtual grid to the top.

However in between the scroll and triggering the click, the virtual grid can hear the scroll, then change which dom row corresponds to which data row (usually in virtual grids, in order that mobile scrolling works you have a window of real dom nodes and upon a scroll you move the data and then the y position of the rows so that the user can continue seamlessly scrolling).

This means that day the grid is rows 1-7 and you get 3.. the trigger is executed on the same dom element that was fetched, but because of the scroll it has become a row containing the data in 5.

Because of race conditions, this usually ends up being intermittent depending on if cypress or the scroll listener wins.

### Desired behavior:

I’d love to be able to disable scroll per command or else entirely so it has to be explicit (better I think for an spa)

@Bkucera this is the scenario I mentioned on the Pr

### LINKED ISSUE #8854: waitForAnimations is not a valid option for .click TS type
### Current behavior
`cy.get('body').click({ waitForAnimations: false })` gets TS error because `waitForAnimations` is not a valid option on the Type even if is a suggested command https://docs.cypress.io/guides/references/error-messages.html#cy-failed-because-the-element-is-currently-animating

### Desired behavior
`cy.get('body').click({ waitForAnimations: false })` will not throw TS error

### Versions
5.3.0 MacOS

### PR DESCRIPTION
**Pull request by lukeapage**
* Closes #8854 
* Closes #4356
* Closes #871 

### User facing changelog
Add a configuration option `scrollBehavior` that controls the viewport position to which an element should be scrolled prior to action commands. Possible values are `'top' 'bottom' 'center' 'nearest' and false`, with `'top'` being the default. `scrollBehavior: false` disables scrolling altogether. `ScrollBehavior` can be specified globally in `cypress.json` or on individual action commands via options.

### Additional details

- options map to those specified for the 'block' property, with 'top' being 'start' and 'bottom' being 'end':  https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView 
Note: this only affects vertical alignment of the element

Docs PR: https://github.com/cypress-io/cypress-documentation/pull/3252

### How has the user experience changed?
No Changes.

### PR Tasks
- [x] rename config value to 'scrollBehavior'
- [x] update docs PR for expanded options
* [x] Have tests been added/updated?

* [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
* [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
* [x] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by lukeapage** [comment]
As I was working I used `waitForAnimations` as a template - but I found that this was not implemented correctly everywhere - some commands don't pass the config option and the typings were incomplete. I have fixed them as I went along, but I'm also happy to remove these fixes.

**Comment by lukeapage** [comment]
app veyor:
```
@packages/desktop-gui: ERROR in ./src/main.scss
@packages/desktop-gui: Module build failed (from ../web-config/node_modules/mini-css-extract-plugin/dist/loader.js):
@packages/desktop-gui: ModuleNotFoundError: Module not found: Error: Can't resolve '../../../node_modules/@reach/dialog/styles.css' in 'C:\projects\cypress\packages\desktop-gui\src'
@packages/desktop-gui:     at C:\projects\cypress\node_modules\webpack-cli\node_modules\webpack\lib\Compilation.js:821:10
@packages/desktop-gui:     at C:\projects\cypress\node_modules\webpack-cli\node_modules\webpack\lib\NormalModuleFactory.js:401:22
@packages/desktop-gui:     at C:\projects\cypress\node_modules\webpack-cli\node_modules\webpack\lib\NormalModuleFactory.js:130:21
@packages/desktop-gui:     at C:\projects\cypress\node_modules\webpack-cli\node_modules\webpack\lib\NormalModuleFactory.js:224:22
@packages/desktop-gui:     at C:\projects\cypress\node_modules\neo-async\async.js:2830:7
@packages/desktop-gui:     at C:\projects\cypress\node_modules\neo-async\async.js:6877:13
@packages/desktop-gui:     at C:\projects\cypress\node_modules\webpack-cli\node_modules\webpack\lib\NormalModuleFactory.js:214:25
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:213:14
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:15:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\UnsafeCachePlugin.js:44:7
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:15:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:27:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\DescriptionFilePlugin.js:67:43
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:27:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\DescriptionFilePlugin.js:67:43
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:16:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\Resolver.js:285:5
@packages/desktop-gui:     at eval (eval at create (C:\projects\cypress\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:15:1)
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\DirectoryExistsPlugin.js:27:15
@packages/desktop-gui:     at C:\projects\cypress\node_modules\enhanced-resolve\lib\CachedInputFileSystem.js:85:15
@packages/desktop-gui:     at processTicksAndRejections (internal/process/task_queues.js:76:11)
@packages/desktop-gui: ERROR in ../ui-components/src/file-opener/editor-picker-modal.tsx
@packages/desktop-gui: Module not found: Error: Can't resolve '@cypress/react-tooltip' in 'C:\projects\cypress\packages\ui-components\src\file-opener'
```

I don't think this is anything to do with me?!

**Comment by lukeapage** [comment]
server performance tests - unrelated to my changes
server-e2e-tests-firefox-6 - e2e video compression passes (headless) [firefox] - passes (headless) [firefox] - unrelated
npm-webpack-preprocessor - this looks unrelated too. 
  1) webpack preprocessor - e2e
       does not call rerun on initial build, but on subsequent builds:
     AssertionError: expected emit to have been called with arguments rerun

File: packages/driver/cypress/integration/commands/actions/click_spec.js
Comment (jennifer-shehane) [inline packages/driver/cypress/integration/commands/actions/click_spec.js:1181]: Can we have a few tests to ensure the scroll actually doesn't happen when passed the option? Just in case something with the `scrollIntoView` function gets changed in the future, to catch that the actual scrolling is broken / not broken.

**Review by jennifer-shehane — DISMISSED** [review]
The rules in actionability apply to several more commands. It may be worthwhile to also test the option in some of those. Up to you. 

I also left a note about testing the actual scroll.

- .click()
- .dblclick()
- .rightclick()
- .type()
- .clear()
- .focus()
- .check()
- .uncheck()
- .select()
- .submit()
- .trigger()

**Comment by lukeapage** [comment]
@jennifer-shehane 
> I also left a note about testing the actual scroll.

Please could you help me out with how you might do this? It was my first thought - and scrollIntoView does scroll even when not needed so I could test this - but the best test would be when the element is hidden - in this case I'd like to assert that the test case fails - Is there an example I can copy of a test that checks that a command will fail?

**Comment by jennifer-shehane** [comment]
Depending on the setup of the test, we would generally check the scroll value of the window itself. So like that the window.scrollY value hasn't changed after the click when it would normally have scrolled: https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollY

If you do what to write an assertion on failure of a test, you can listen to the `cy.on('fail')...` event listener and makes assertions in there, then fire the `done()` callback when you want the test to be considered done. Similar to this test: https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/integration/commands/actions/scroll_spec.js#L443:L443

File: packages/driver/cypress/integration/commands/actions/click_spec.js
Comment (lukeapage) [reply packages/driver/cypress/integration/commands/actions/click_spec.js:1181]: added tests that assert it is not scrolled.

**Comment by lukeapage** [comment]
@jennifer-shehane added tests

**Comment by lukeapage** [comment]
There are some failures in tests I added in Firefox and one in chrome that I need to fix - just noticed in the ci logs

**Comment by lukeapage** [comment]
(Sure they didn’t happen locallly)

**Comment by lukeapage** [comment]
Tests should be fixed now. They ran on their own but not when run with other tests - the issue is that those specs do not reload the page every test and the scroll position was persisting between tests. Adding it in the beforeEach broke a test and I wasn't able to fix that test because it had a deep assertion with a number that differed between firefox and chrome ;- I don't want to have to refactor all the tests for this PR so I kept my test changes localised.

**Comment by jennifer-shehane** [comment]
I think there may be some hesitation with introducing this option, but I'd like to advocate for adding this also. Some examples where this will come in handy as a tool for when bugs arise, to implement a workaround or your own scrolling strategy.

- #4233 - with the `scrollToElement` option passed to the click here, this issue can be avoided since our scroll algorithm is incorrect. So this offers a workaround. 
- #2302 - this also passes when passing `scrollToElement` option. Can be offered as workaround.

I manually checked that this works as intended for `.click()`.

**Comment by lukeapage** [comment]
Is this just waiting for a review from @Bkucera or are you having an internal discussion about whether to accept it or not?

**Comment by jennifer-shehane** [comment]
Yes, awaiting @Bkucera's review when he has time.

**Comment by lukeapage** [comment]
Is @Bkucera aware this is waiting on him? Its been 10 days now, it would be good to know even roughly, when he will have time

**Comment by jennifer-shehane** [comment]
@lukeapage Yes, he is going to try to review it as soon as possible, while also considering his other work.

**Comment by lukeapage** [comment]
Can someone else review it?

**Comment by Bkucera** [comment]
@lukeapage I'll take a look this week. thanks for your patience

**Comment by Bkucera** [comment]
expanded the option to be able to specify scroll behavior as requested in #871. @lukeapage let me know what you think. We can get this out in the next feature release after I get one more set of eyes on the API

**Comment by lukeapage** [comment]
Thanks - no objection from me for that. There’s a docs Pr linked at the top that needs updating with the additional functionality.

**Review by Bkucera — APPROVED** [review]

**Pull request merged**
